### PR TITLE
Test WPJ leave 1561081

### PR DESCRIPTION
**Pull request by wainaina**
Automated Test Case (https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1561081)

### CODE REVIEW

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+     * and confirms the deivce's registration / de-registration status.
+     *
+     * @param deviceID - The device Id obtained from performDeviceRegistration
+     * @param username - LabApi username
+     * @param password - LabApi password
+     * @param isDeviceRegistered - toggle to enable re-usability of this method across the two states
+     *                           this device maybe in, ie REGISTERED / UNREGISTERED.
+     * @throws Throwable
+     */
+    private void getDeviceIdFromMyAccount(String deviceID, String username, String password, boolean isDeviceRegistered) throws  Throwable{
```
Comment (iamgusain) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:162]: Should we make this a helper method (UIAutomationUtilities), so other test cases can also use this.

**Review by iamgusain — APPROVED** [review]
LGTM

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
+//  Copyright (c) Microsoft Corporation.
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:1]: Please put complete LICENSE

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+public class TestCase1561081 extends AbstractMsalBrokerTest {
+
+    final String MY_ACCOUNT_MICROSOFT_URL = "https://myaccount.microsoft.com/";
+
+    @Test
+    public void test_1561081() throws Throwable {
+        final String username = mLabAccount.getUsername();
+        final String password = mLabAccount.getPassword();
+
+        final BrokerMicrosoftAuthenticator brokerMicrosoftAuthenticator = (BrokerMicrosoftAuthenticator) mBroker;
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:68]: How do you know this is indeed the Broker being used? Line 37 declares two brokers as supported and if this is different then you get a class cast exception

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+        //perform device registration
+        brokerMicrosoftAuthenticator.performDeviceRegistration(username, password);
+
+        final MsalSdk msalSdk = new MsalSdk();
+
+        MsalAuthTestParams authTestParams_firstTry = MsalAuthTestParams.builder()
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java]:
```suggestion
        final MsalAuthTestParams authTestParams_firstTry = MsalAuthTestParams.builder()
```

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                .activity(mActivity)
+                .loginHint(username)
+                .scopes(Arrays.asList(mScopes))
+                .promptParameter(Prompt.LOGIN)
+                .msalConfigResourceId(getConfigFileResourceId())
+                .build();
+
+        //AT interactive acquisition.
+        MsalAuthResult authResult = msalSdk.acquireTokenInteractive(authTestParams_firstTry, new OnInteractionRequired() {
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java]:
```suggestion
        final MsalAuthResult authResult = msalSdk.acquireTokenInteractive(authTestParams_firstTry, new OnInteractionRequired() {
```

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+                new MicrosoftStsPromptHandler(promptHandlerParameters)
+                        .handlePrompt(username, password);
+            }
+        }, TokenRequestTimeout.MEDIUM);
+
+        authResult.assertSuccess();
+
+        //extract the device id claim from the access token.
+        String deviceId = (String) IDToken.parseJWT(authResult.getAccessToken()).get("deviceid");
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java]: I suggest we change this to rather use the JWT Parser class that the rest of the tests are using

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+        //this gets the deviceId from the Ui and matches it to the deviceID obtained from the AT
+        getDeviceIdFromMyAccount(deviceId, username, password, true);
+
+        //init the release candidate BrokerHost in order to run wpj Leave
+        BrokerHost brokerHost = new BrokerHost(BrokerHost.BROKER_HOST_APK_RC);
+        if(brokerHost.isInstalled()){
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:109]: In what cases would it already be installed?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java (hunk 1 above)
Comment (shahzaibj) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:162]: We should

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java (hunk 3 above)
Comment (wainaina) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:68]: I have removed BrokerHost as a supported broker.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java (hunk 1 above)
Comment (wainaina) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:162]: We already have the get device ID in the BrokerHost class that does the same thing. This method just uses the slightly longer approach to get it and compares it against the deviceId in the token. We can achieve the same using the brokerHost.getdeviceId() method.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                        .build();
+
+                new MicrosoftStsPromptHandler(promptHandlerParameters)
+                        .handlePrompt(username, password);
+            }
+        }, TokenRequestTimeout.MEDIUM);
+
+        authResult2.assertSuccess();
+
+        //wait for two minutes to ensure the changes are made.
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:144]: What changes?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+     * and confirms the deivce's registration / de-registration status.
+     *
+     * @param deviceID - The device Id obtained from performDeviceRegistration
+     * @param username - LabApi username
+     * @param password - LabApi password
+     * @param isDeviceRegistered - toggle to enable re-usability of this method across the two states
+     *                           this device maybe in, ie REGISTERED / UNREGISTERED.
+     * @throws Throwable
+     */
+    private void getDeviceIdFromMyAccount(String deviceID, String username, String password, boolean isDeviceRegistered) throws  Throwable{
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:162]: method can be renamed to `getAndConfirmDeviceIdFromMyAccount`

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+                        .sessionExpected(false)
+                        .consentPageExpected(false)
+                        .build();
+
+                new MicrosoftStsPromptHandler(promptHandlerParameters)
+                        .handlePrompt(username, password);
+            }
+        }, TokenRequestTimeout.MEDIUM);
+
+        authResult2.assertSuccess();
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:142]: I think per the test, we want to again use the JWT Parser here and confirm that now the token does NOT contain `deviceid`

**Review by shahzaibj — APPROVED** [review]
I'm signing off but I left some comments that I suggest you consider prior to merging.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+        //wait for two minutes to ensure the changes are made.
+        Thread.sleep(TimeUnit.MINUTES.toMinutes(2));
```
Comment (bktowett) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:145]: What is the rationale for choosing two minutes?

Comment (wainaina) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/TestCase1561081.java:145]: To give it more than enough time to have the changes propagated in order to have a consistent response on getting the device Id from account.

**Pull request merged**